pika/channel.py

```diff
@@ -944,32 +1015,80 @@ def _on_cancelok(self, method_frame):
     def _on_close(self, method_frame):
         """Handle the case where our channel has been closed for us
 
-        :param pika.frame.Method method_frame: The close frame
+        :param pika.frame.Method method_frame: Method frame with Channel.Close
+            method
 
         """
-        LOGGER.info('%s', method_frame)
-        LOGGER.warning('Received remote Channel.Close (%s): %r on channel %s',
+        LOGGER.warning('Received remote Channel.Close (%s): %r on %s',
                        method_frame.method.reply_code,
                        method_frame.method.reply_text,
                        self)
-        if self.connection.is_open:
-            self._send_method(spec.Channel.CloseOk())
-        self._set_state(self.CLOSED)
-        self.callbacks.process(self.channel_number, '_on_channel_close', self,
-                               self, method_frame.method.reply_code,
-                               method_frame.method.reply_text)
-        self._cleanup()
+
+        # AMQP 0.9.1 requires CloseOk response to Channel.Close; Note, we should
+        # not be called when connection is closed
+        self._send_method(spec.Channel.CloseOk())
```

Comment:
@gmr, regarding
> Shouldn't this come after the check on 1031?

No, it should not. Per AMQP 0.9.1, each peer must respond to the other peer's `Channel.Close`. This is what this specific compound scenario looks like:

pika: sends `Channel.Close` to broker

broker: sends `Channel.Close` to pika client (e.g., in response to error from prior client request on this channel)

pika gets `Channel.Close` from broker and is required to reply to broker  with `Channel.CloseOk` per AMQP 0.9.1.

broker gets `Channel.Close` from pika client and is required to reply with `Channel.Close.Ok` (even if the broker already sent `Channel.Close` for some reason) 

Since pika client sent `Channel.Close` earlier, it should expect to get for `Channel.CloseOk` from the broker per AMQP 0.9.1.

That's why `self._send_method(spec.Channel.CloseOk())` needs to come before the check `if self.is_closing` on 1031